wz_strategic_plugin/src/wz_strategic_plugin.cpp

```diff
@@ -0,0 +1,508 @@
+
+  ROS_DEBUG("traffic_light_down_track %f", traffic_light_down_track);
+
+  auto current_light_state_optional = nearest_traffic_light->getState();
+
+  if (!validLightState(current_light_state_optional, ros::Time::now()))
+    return;
+
+  if (current_light_state_optional.get() == lanelet::CarmaTrafficLightState::PERMISSIVE_MOVEMENT_ALLOWED)
+  {
+    transition_table_.signal(TransitEvent::RED_TO_GREEN_LIGHT);  // If the light is green send the light transition
+                                                                 // signal
+    return;
+  }
+
+  // If the light is not green then continue waiting by creating a stop and wait maneuver ontop of the vehicle
+  resp.new_plan.maneuvers.push_back(
+      composeStopAndWaitManeuverMessage(current_state.downtrack - 10.0, traffic_light_down_track, current_state.speed,
+                                        current_state.lane_id, current_state.lane_id, current_state.stamp,
+                                        current_state.stamp + ros::Duration(config_.min_maneuver_planning_period) ));
+}
+
+void WzStrategicPlugin::planWhenDEPARTING(const cav_srvs::PlanManeuversRequest& req,
+                                          cav_srvs::PlanManeuversResponse& resp, const VehicleState& current_state,
+                                          double intersection_end_downtrack, double intersection_speed_limit)
+{
+  if (current_state.downtrack > intersection_end_downtrack)
+  {
+    transition_table_.signal(TransitEvent::INTERSECTION_EXIT);  // If we are past the most recent
+    return;
+  }
+
+  ros::Time intersection_exit_time =
+      current_state.stamp +
+      ros::Duration((intersection_end_downtrack - current_state.downtrack) / intersection_speed_limit);
+
+  // Identify the lanelets which will be crossed by approach maneuvers lane follow maneuver
+  std::vector<lanelet::ConstLanelet> crossed_lanelets =
+      getLaneletsBetweenWithException(current_state.downtrack, intersection_end_downtrack, true, false);
+
+  // Compose intersection transit maneuver
+  resp.new_plan.maneuvers.push_back(composeIntersectionTransitMessage(
+      current_state.downtrack, intersection_end_downtrack, intersection_speed_limit, intersection_speed_limit,
+      current_state.stamp, intersection_exit_time, crossed_lanelets.front().id(), crossed_lanelets.back().id()));
+}
+
+bool WzStrategicPlugin::planManeuverCb(cav_srvs::PlanManeuversRequest& req, cav_srvs::PlanManeuversResponse& resp)
+{
+  if (!wm_->getRoute())
+  {
+    ROS_ERROR_STREAM("Could not plan maneuvers as route was not available");
+    return true;
+  }
+
+  ROS_DEBUG("Finding car information");
+
+  // Extract vehicle data from request
+  VehicleState current_state = extractInitialState(req);
+
+  // Get current traffic light information
+  ROS_DEBUG("\n\nFinding traffic_light information");
+
+  auto traffic_list = wm_->getLightsAlongRoute({ req.veh_x, req.veh_y });
```

Comment:
break out of which loop? 
If you mean the state transition loop I would rather allow the states to stop transitioning before breaking out to ensure the current state is consistent with the planning environment on planning completion.

Reply:
I meant the loop in the getLightsAlongRoute function in carma world model, instead of checking the complete shortest path, we should be only checking till the current downtrack.

Reply:
Looks like the intention of the function is to return a sorted list (by downtrack distance) of traffic lights on the route ahead of us, which it is doing